### ISSUE TITLE
Monitor Settings: Unplugging external monitor makes it impossible to switch off extended desktop

### ISSUE THREAD

**Issue by DrLuny** [lxde/lxqt#600]
When I unplug my VGA monitor from my laptop without first changing to LVDI-only in the quick options, the quick options page no longer appears and I can't use it to eliminate the extended desktop(open windows remain on the side of the desktop the monitor was displaying and I can drag things off my screen).  Applying the settings I can set at the settings tab, which no longer sees the VGA display, does not remove the extra desktop space.  I don't think detecting the monitor being unplugged and changing the configuration is necessary or wise, but it should be trivial to remove unconnected displays from the configuration when I hit the apply button. Another option might by tying the display of the quick options tab to the current configuration rather than the monitors detected, but that could raise issues if users modify the script manually and would probably be harder to implement.

I'd also like to mention that the save and apply options are a bit confusing.  I really liked the xrandr gui that comes with openbox and the approach of being able to save and load configuration scripts in a file chooser. The ability to drag the screens to different locations was very nice too(I don't have my second monitor with me at the moment so I forget whether that's included or not). It also feels like its buried a level too deep in the menu. The Quick Options tab is a nice feature, though, because regardless of the screen configurer's features, being able to quickly plug in a monitor and have it display appropriately is very convenient.

**Comment by selairi** [comment]
I will try to fix this bug. Thanks for your report. I'm working in a
new interface, with saving options.

Thanks

2015-04-17 19:25 GMT+02:00, DrLuny <notifications@github.com>:

**Comment by selairi** [comment]
I have just pushed to Github a possible solution to remove the extra 
desktop space. It is in:

https://github.com/lxde/lxqt-config

X11 should calculate framebuffer size by itself. It usually does it 
bigger. I forgot to make it smaller. Sorry

I also include some screenshots of new interface. Saving options is in 
other tab. When I finish writing it, you will be able to save configs 
and load them when compatible hardware is detected.

Thanks

**Comment by palinek** [comment]
@selairi there is a work in progress on migration to `KScreen` of the whole `lxqt-config-monitor`. See #502.

**Comment by selairi** [comment]
I have already read that thread. I have seen your work in:

https://github.com/palinek/lxqt-config/tree/switch-to-kscreen

I was thinking of "stealing" part of your code and implement kscreen
as backend.I think it is not a good idea to remove support for
backend. Wayland is not clear in screen settings and if you see WRandR
(http://lists.freedesktop.org/archives/wayland-devel/2014-March/013843.html),
screen config will depend on compositor.

If lxqt-config-monitor is moved to Kscreen, lxqt only will be able to
change screen size under Kwin.

I prefer keep an abstract layer and change backend if we need it.

I have started to rewrite backend interface to improve it.

Thanks





2015-04-20 9:14 GMT+02:00, palinek <notifications@github.com>:

**Comment by paulolieuthier** [comment]
@selairi AFAIK KScreen works on other window managers like Openbox as well.

**Comment by jleclanche** [comment]
kscreen works just fine on any window manager... I don't see why kwin was
brought into it.

Wayland doesn't have a screen management protocol yet. The KScreen dev
confirmed there'll be support for it when there is.

I'm -2 on keeping the current xrandr interface. It's really nasty and
there's no reason to keep it if we have kscreen support, which gives us
more than what xrandr does.

J. Leclanche

**Comment by selairi** [comment]
Does kscreen work with other wayland compositors? I would prefer use
kscreen as backend and keep xrandr backend.

xrandr works and is lighter than kscreen. User can select kscreen or
xrandr. And finally lxqt-config-monitor can be easily adapted to
wayland.



2015-04-21 17:33 GMT+02:00, Paulo Lieuthier <notifications@github.com>:

**Comment by jleclanche** [comment]
There's no such thing as a library, or protocol, for screen management on
wayland. Xrandr certainly doesn't work.

The GNOME environment use a nonstandard port of their xrandr dbus
interface, but it's all done in-compositor, no library at work.

This is an issue that we can't fix right now. I will follow it up, but we
shouldnt worry about it yet.

J. Leclanche

**Comment by selairi** [comment]
>
> I'm -2 on keeping the current xrandr interface. It's really nasty and
> there's no reason to keep it if we have kscreen support, which gives us
> more than what xrandr does.
>

Xrandr command line is fast and uses low memory. You can speed up your 
desktop starting. Kscreen interface will made it slower. A Ten-years-old 
computer doesn't need a heavy interface.

The current xrandr interface is simple and bugs can be easily found. We 
can change the backend and we can reuse the ui interface. And it is 
possible use both backends.

**Comment by jleclanche** [comment]
On 21 April 2015 at 23:07, P.L. Lucas <notifications@github.com> wrote:

>
> Xrandr command line is fast and uses low memory. You can speed up your
> desktop starting. Kscreen interface will made it slower. A Ten-years-old
> computer doesn't need a heavy interface.


You're making a lot of wrong assumptions.

J. Leclanche

**Comment by selairi** [comment]
>
> >
> > Xrandr command line is fast and uses low memory. You can speed up your
> > desktop starting. Kscreen interface will made it slower. A Ten-years-old
> > computer doesn't need a heavy interface.
>
>
> You're making a lot of wrong assumptions.
>
>


>
Since I have been working hard for a very long time on this topic, I 
consider it is not a proper answer to say just: "You are making a lot of 
wrong assumptions" without a logic explanation.

**Comment by jleclanche** [comment]
libkscreen communicates directly with x11. The xrandr backend goes through
a pipe. We ran the numbers when I was working on it and using kscreen is
far faster than the current xrandr backend - and there is no reason it
shouldn't be. It also creates no new dependency other than libkscreen
itself, which is very small and not in any way "heavy" (<700KB on disk).
And, it has a much nicer API than the xrandr command line, doesn't involve
popen nastiness and what not.
It also offloads "backend maintenance" outside of lxqt. We don't want to be
maintaining backends for a tool that depends on a library that already has
a backend system.

J. Leclanche

**Comment by selairi** [comment]
Thanks for your reply.

lxqt-config-monitor saves your config in a simple .desktop file. This 
.desktop file contains xrandr command to set your screen config at 
starting. It is not necessary load screen config at desktop init (no 
pipes at start).

The main goal is speed up your desktop starting.

If community prefer kscreen interface, I will help on it. But the fisrt 
idea was another.



> libkscreen communicates directly with x11. The xrandr backend goes through
> a pipe. We ran the numbers when I was working on it and using kscreen is
> far faster than the current xrandr backend - and there is no reason it
> shouldn't be. It also creates no new dependency other than libkscreen
> itself, which is very small and not in any way "heavy" (<700KB on disk).
> And, it has a much nicer API than the xrandr command line, doesn't involve
> popen nastiness and what not.
> It also offloads "backend maintenance" outside of lxqt. We don't want 
> to be
> maintaining backends for a tool that depends on a library that already has
> a backend system.
>

**Comment by jleclanche** [comment]
The desktop file issue is different. We discussed it on IRC when this was brought up but no concrete solution was proposed. Ideally we should save these things in the xorg config rather than run xrandr commands at startup, but that's another can of worms.

**Comment by Vladimir-csp** [comment]
Some user-level daemon is needed anyway to handle display hotplug events and apply configurations accordingly. It would also replace xrandr command on startup

**Comment by jleclanche** [comment]
Yes. We discussed bringing the kscreen daemon but it is massive, far too heavy for our task. Ideally we're looking for a separate, third party component which would use libkscreen. This is open to contributors.

**Comment by palinek** [comment]
> We discussed bringing the kscreen daemon but it is massive, far too heavy for our task. 

Just recently I've implemented listening for new input devices by `libudev` into `lxqt-session` (lxde/lxqt-session@33a5f92f4fbdbafb24c38de90adf551065447b95) and IMHO we could easily use that also to trigger any screen dependent operation.
